<a href="https://colab.research.google.com/github/abdishifa234/DS-Unit-1-Sprint-2-Statistics/blob/master/Module2/Copy_of_LS_DS_122_Chi2_Tests_Assignment.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## Assignment - Practice Chi^2 Tests

Use the following dataset relating to math scores of students in two different Portugese schools:

<https://archive.ics.uci.edu/ml/datasets/Student+Performance>

### 1) Load the dataset specifically relating to math scores as a new dataframe.

There are two datasets in the `student.zip` file, make sure you use `student-mat.csv`.


In [ ]:
# YOUR WORK HERE
! wget https://archive.ics.uci.edu/ml/machine-learning-databases/00320/student.zip

--2020-07-25 23:18:34--  https://archive.ics.uci.edu/ml/machine-learning-databases/00320/student.zip
Resolving archive.ics.uci.edu (archive.ics.uci.edu)... 128.195.10.252
Connecting to archive.ics.uci.edu (archive.ics.uci.edu)|128.195.10.252|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 20478 (20K) [application/x-httpd-php]
Saving to: ‘student.zip.3’

student.zip.3       100%[===================>]  20.00K  --.-KB/s    in 0.04s   

2020-07-25 23:18:34 (536 KB/s) - ‘student.zip.3’ saved [20478/20478]



In [ ]:
import pandas as pd
import numpy as np
from scipy import stats
df=pd.read_csv('student-mat.csv',sep=';')
print(df.shape)
df.head()

(395, 33)


school sex  age address famsize Pstatus  ...  Walc  health absences  G1  G2  G3
0     GP   F   18       U     GT3       A  ...     1       3        6   5   6   6
1     GP   F   17       U     GT3       T  ...     1       3        4   5   5   6
2     GP   F   15       U     LE3       T  ...     3       3       10   7   8  10
3     GP   F   15       U     GT3       T  ...     1       5        2  15  14  15
4     GP   F   16       U     GT3       T  ...     2       5        4   6  10  10

[5 rows x 33 columns]

### 2) Use Chi^2 tests and `stats.chi2_contingency()` to identify:
 - Two pairs of variables that are dependent (are associated with one another).
 - Two pairs of variables that are independent (have no significant relationship).

Does it make intuitive sense why the variables in these pairs might or might not show a relationship to one another? 


OK lets number of school absence and failures and do chi square tests
#Assumptions
1.Null Hyphothesis:There is no association between the number of school absence and failures.This is to say that the two variables are independent

2.Alternate Hyphothesis:There is association between the number of school absence and failures.This is to say that the two variables are dependant

3.Confidence Level:95%



In [ ]:
# Lets check the the frequencies of failures
df['failures'].value_counts().sort_index()

0    312
1     50
2     17
3     16
Name: failures, dtype: int64

In [ ]:
# Lets check frequencies of the number of absences
df['absences'].value_counts().sort_index().head()

0    115
1      3
2     65
3      8
4     53
Name: absences, dtype: int64

In [ ]:
#Ok lets do crosstans or contigency table of the two variables
contigency=pd.crosstab(df['failures'],df['absences'])
contigency

absences  0   1   2   3   4   5   6   7   ...  26  28  30  38  40  54  56  75
failures                                  ...                                
0         89   3  58   6  47   4  25   6  ...   1   0   1   0   0   1   1   1
1         14   0   4   1   4   0   3   1  ...   0   1   0   1   1   0   0   0
2          5   0   1   0   2   0   2   0  ...   0   0   0   0   0   0   0   0
3          7   0   2   1   0   1   1   0  ...   0   0   0   0   0   0   0   0

[4 rows x 34 columns]

In [ ]:
#Lets do chi square test
stats.chi2_contingency(contigency)

(122.18899006084003,
 0.05694057065252934,
 99,
 array([[9.08354430e+01, 2.36962025e+00, 5.13417722e+01, 6.31898734e+00,
         4.18632911e+01, 3.94936709e+00, 2.44860759e+01, 5.52911392e+00,
         1.73772152e+01, 2.36962025e+00, 1.34278481e+01, 2.36962025e+00,
         9.47848101e+00, 2.36962025e+00, 9.47848101e+00, 2.36962025e+00,
         5.52911392e+00, 7.89873418e-01, 3.94936709e+00, 7.89873418e-01,
         3.15949367e+00, 7.89873418e-01, 2.36962025e+00, 7.89873418e-01,
         7.89873418e-01, 7.89873418e-01, 7.89873418e-01, 7.89873418e-01,
         7.89873418e-01, 7.89873418e-01, 7.89873418e-01, 7.89873418e-01,
         7.89873418e-01, 7.89873418e-01],
        [1.45569620e+01, 3.79746835e-01, 8.22784810e+00, 1.01265823e+00,
         6.70886076e+00, 6.32911392e-01, 3.92405063e+00, 8.86075949e-01,
         2.78481013e+00, 3.79746835e-01, 2.15189873e+00, 3.79746835e-01,
         1.51898734e+00, 3.79746835e-01, 1.51898734e+00, 3.79746835e-01,
         8.86075949e-01, 1.2658227

#Analysis 
As you can see from the chi square result above
#chi2  is  122.18899
#p-value is  0.05694057065252934
#Degrees of fredom is    99
#Expected frequency table

#Conclusion
The decsion rule for the chi sqare test is just the same as t-test which is  based on the p-value but the interpretation is different.But as you can see from the results of the chi square test above the P-value of   0.05694057065252934 is grater than 0.05 which force us not to reject the null hyphothesis that says "There is no association b/n the number of school absence and school failurs eventhough in reality there is strong association b/n school absence and exam failures.


OK lets consider another variables  extra paid classes within the course subject and weekly study time and do chi square tests

Assumptions
1.Null Hyphothesis:There is no association between student's extra paid classes and study tiime.This is to say that the two variables are independent

2.Alternative Hyphothesis:There is an association between student's extra paid classes and study tiime.This is to say that the two variables are dependent
3.Confidence Level:95%

In [ ]:
# df.paid.replace(('yes', 'no'), (1, 0), inplace=True)
df.head()

school sex  age address famsize Pstatus  ...  Walc  health absences  G1  G2  G3
0     GP   F   18       U     GT3       A  ...     1       3        6   5   6   6
1     GP   F   17       U     GT3       T  ...     1       3        4   5   5   6
2     GP   F   15       U     LE3       T  ...     3       3       10   7   8  10
3     GP   F   15       U     GT3       T  ...     1       5        2  15  14  15
4     GP   F   16       U     GT3       T  ...     2       5        4   6  10  10

[5 rows x 33 columns]

In [ ]:
#Ok lets do crosstans or contigency table of the two variables
contigency2=pd.crosstab(df['paid'],df['studytime'])
contigency2

studytime   1   2   3   4
paid                     
0          73  99  31  11
1          32  99  34  16

In [ ]:
#Lets do chi square test
stats.chi2_contingency(contigency2)

(14.417578913306178,
 0.002388496856798663,
 3,
 array([[ 56.88607595, 107.27088608,  35.21518987,  14.6278481 ],
        [ 48.11392405,  90.72911392,  29.78481013,  12.3721519 ]]))

#Analysis 
As you can see from the chi square result above
#chi2  is  14.4175789
#p-value is  0.00238849
#Degrees of fredom is    3
#Expected frequency table

#Conclusion
 As you can see from the results of the chi square test above, the P-value of   0.00238849 is less than 0.05 which leads to reject the null hyphothesis that says "There is no association b/n the extra paid class and study time and the two variables are dependant.

### 3) Use NumPy to perform your own chi^2 test "from scratch" 

Pick any of the chi^2 tests that you ran in part 2 and try them on your own. You should get the same results that Scipy got for all four values returned from `chi2_contingency()`

In [ ]:
# Lets take school absence and failures that we considerd above in part one
# make the contigency table
contigency_table=pd.crosstab(df['failures'],df['absences'],margins=True)
contigency_table


absences    0  1   2  3   4  5   6  7   8  ...  26  28  30  38  40  54  56  75  All
failures                                   ...                                     
0          89  3  58  6  47  4  25  6  17  ...   1   0   1   0   0   1   1   1  312
1          14  0   4  1   4  0   3  1   2  ...   0   1   0   1   1   0   0   0   50
2           5  0   1  0   2  0   2  0   2  ...   0   0   0   0   0   0   0   0   17
3           7  0   2  1   0  1   1  0   1  ...   0   0   0   0   0   0   0   0   16
All       115  3  65  8  53  5  31  7  22  ...   1   1   1   1   1   1   1   1  395

[5 rows x 35 columns]

## Chi-Squared Statistic with Numpy

\begin{align}
\chi^2 = \sum \frac{(observed_{i}-expected_{i})^2}{(expected_{i})}
\end{align}

For the $observed$ values we will just use a version of our contingency table without the margins as a numpy array. In this way, if our observed values array and our expected values array are the same shape, then we can subtract them and divide them directly which makes the calculations a lot cleaner. No for loops!

In [ ]:
row_sums = contingency_table.iloc[0:3, 34].values
col_sums = contingency_table.iloc[2, 0:6].values

print(row_sums)
print(col_sums)

NameError: ignored

## Expected Value Calculation
\begin{align}
expected_{i,j} =\frac{(row_{i} \text{total})(column_{j} \text{total}) }{(\text{total observations})}  
\end{align}

In [ ]:
expected = []
for i in range(len(row_sums)):
    expected_row = []
    for column in col_sums:
        expected_val = column*row_sums[i]/total
        expected_row.append(expected_val)
    expected.append(expected_row)
    
expected = np.array(expected)
print(expected.shape)  
print(expected)

NameError: ignored

In [ ]:
observed = pd.crosstab(df['failures'], df['absences']).values
print(observed.shape)
observed

(4, 34)


array([[89,  3, 58,  6, 47,  4, 25,  6, 17,  3, 15,  2,  9,  2,  8,  0,
         3,  1,  4,  0,  2,  1,  1,  1,  0,  0,  1,  0,  1,  0,  0,  1,
         1,  1],
       [14,  0,  4,  1,  4,  0,  3,  1,  2,  0,  1,  0,  3,  1,  2,  2,
         2,  0,  1,  1,  2,  0,  1,  0,  1,  1,  0,  1,  0,  1,  1,  0,
         0,  0],
       [ 5,  0,  1,  0,  2,  0,  2,  0,  2,  0,  0,  1,  0,  0,  2,  1,
         0,  0,  0,  0,  0,  0,  1,  0,  0,  0,  0,  0,  0,  0,  0,  0,
         0,  0],
       [ 7,  0,  2,  1,  0,  1,  1,  0,  1,  0,  1,  0,  0,  0,  0,  0,
         2,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,
         0,  0]])

In [ ]:
# Array broadcasting will work with numpy arrays but not python lists
chi_squared = ((observed - expected)**2/(expected)).sum() 
print(f"Chi-Squared: {chi_squared}")

In [ ]:
# Degrees of Freedom of a Chi-squared test

#degrees_of_freedom = (num_rows - 1)(num_columns - 1)


# Calculate Degrees of Freedom
dof = (len(row_sums)-1)*(len(col_sums)-1)
print(f"Degrees of Freedom: {dof}") 

## Stretch goals:

### 1. Refactor your code so it is elegant, readable, and holds reusable code in functions.

In [ ]:
# YOUR WORK HERE



### 2. Check For Understanding - Study and write your own explanations/definitions for these topics:

- What is a sample "estimate" in statistics?


#A statistic is an estimator of some parameter in a population. For example:The sample standard deviation (s) is a point estimate of the population standard deviation (σ). The sample mean (̄x) is a point estimate of the population mean, μ.The sample variance (s2 is a point estimate of the population variance (σ2).

- What are hypothesis test? How are they useful?

#Hypothesis testing refers to the process of making inferences or educated guesses about a particular parameter. This can either be done using statistics and sample data, or it can be done on the basis of an uncontrolled observational study. 

- What is a "null hypothesis?"

#A Null hyphothesis is ahyphothesis of no difference or it is a boring hyphothesis hahahaha

#Null hypothesis (often denoted H0,[1]) is a general statement or default position that there is no relationship between two measured phenomena or no association among groups


- What is a p-value and what does it represent?

#The p-value is the probability of obtaining results as extreme as the observed results of a statistical hypothesis test, assuming that the null hypothesis is correct. ... A smaller p-value means that there is stronger evidence in favor of the alternative hypothesis.

- What does it mean for something to be "statistically significant?"

A result has statistical significance when it is very unlikely to have occurred given the null  hypothesis. More precisely, a study's defined significance level, denoted by , is the probability of the study rejecting the null hypothesis, given that the null hypothesis was assumed to be true; and the p-value of a result,  is the probability of obtaining a result at least as extreme, given that the null hypothesis is true. The result is statistically significant, by the standards of the study, when  p-value is less than alpha .

#to determine if statistical significance we consider the p-value and the confidence interval.

- What is a test statistic and how does it relate to a p-value?

# They are inversely proportional to one another

- What are "degrees of freedom" and how are they calculated in a 1-sample, 2-sample, and chi^2 test for independence?

#one sample
Degrees of Freedom: What are they?
Share on

Degrees of freedom are used in hypothesis testing.

Contents (click to skip to that section):

What are Degrees of Freedom?

#Degrees of freedom of an estimate is the number of independent pieces of information that went into calculating the estimate. 
#one sample Tset
Degrees of Freedom = n – 1
#Two sample Test
Degrees of Freedom (Two Samples): (N1 + N2) – 2.

## Resources

- [Interactive visualize the Chi-Squared test](https://homepage.divms.uiowa.edu/~mbognar/applets/chisq.html)
- [Calculation of Chi-Squared test statistic](https://en.wikipedia.org/wiki/Pearson%27s_chi-squared_test)
- [Visualization of a confidence interval generated by R code](https://commons.wikimedia.org/wiki/File:Confidence-interval.svg)
- [Expected value of a squared standard normal](https://math.stackexchange.com/questions/264061/expected-value-calculation-for-squared-normal-distribution) (it's 1 - which is why the expected value of a Chi-Squared with $n$ degrees of freedom is $n$, as it's the sum of $n$ squared standard normals)